### ISSUE TITLE
Tests fail for arkode/CXX_parallel/ark_heat2D.cpp example

### ISSUE THREAD

**Issue by idontgetoutmuch** [LLNL/sundials#26]
If I build `SUNDIALS` as given [here](https://github.com/LLNL/sundials/blob/master/examples/arkode/CXX_parallel/README) and encoded as a nix derivation thus

```
{ stdenv
, cmake
, fetchurl
, openmpi
, llvmPackages
, python
, liblapack
, gfortran
, lapackSupport ? true }:

let liblapackShared = liblapack.override {
  shared = true;
};
    openmp = llvmPackages.openmp;

in stdenv.mkDerivation rec {
  pname = "sundials";
  version = "5.0.0";

  buildInputs = [ python openmpi openmp ] ++ stdenv.lib.optionals (lapackSupport) [ gfortran ];
  nativeBuildInputs = [ cmake openmpi openmp ];

  src = fetchurl {
    url = "https://computing.llnl.gov/projects/${pname}/download/${pname}-${version}.tar.gz";
    sha256 = "1lvx5pddjxgyr8kqlira36kxckz7nxwc8xilzfyx0hf607n42l9l";
  };

  patches = [
    (fetchurl {
      # https://github.com/LLNL/sundials/pull/19
      url = "https://github.com/LLNL/sundials/commit/1350421eab6c5ab479de5eccf6af2dcad1eddf30.patch";
      sha256 = "0g67lixp9m85fqpb9rzz1hl1z8ibdg0ldwq5z6flj5zl8a7cw52l";
    })
    (fetchurl {
      # https://github.com/LLNL/sundials/pull/20
      url = "https://github.com/LLNL/sundials/pull/20/commits/2d951bbe1ff7842fcd0dafa28c61b0aa94015f66.patch";
      sha256 = "0lcr6m4lk14yqrxah4rdscpczny5l7m1zpfsjh8bgspadfsgk512";
    })
  ];

  cmakeFlags = [
    "-DEXAMPLES_INSTALL_PATH=${placeholder "out"}/share/examples"
  ] ++ stdenv.lib.optionals (lapackSupport) [

    "-DSUNDIALS_INDEX_SIZE=64"
    "-DMPI_ENABLE=ON"
    "-DMPI_C_COMPILER=${openmpi}/bin/mpicc"
    "-DMPI_CXX_COMPILER=${openmpi}/bin/mpicxx"
    "-DMPIEXEC_EXECUTABLE=${openmpi}/bin/mpirun"

    "-DOPENMP_ENABLE=ON"

    "-DBUILD_SHARED_LIBS=OFF"
    "-DBUILD_STATIC_LIBS=ON"

    "-DBUILD_CVODE=ON"
    "-DBUILD_CVODES=OFF"
    "-DBUILD_IDA=OFF"
    "-DBUILD_IDAS=OFF"
    "-DBUILD_ARKODE=ON"
    "-DBUILD_KINSOL=OFF"
    "-DBUILD_TESTING=ON"
    "-DSUNDIALS_DEVTESTS=ON"
    "-DEXAMPLES_ENABLE_CXX=ON"

    "-DLAPACK_ENABLE=OFF"
    "-DLAPACK_LIBRARIES=${liblapackShared}/lib/liblapack${stdenv.hostPlatform.extensions.sharedLibrary};${liblapackShared}/lib/libblas${stdenv.hostPlatform.extensions.sharedLibrary}"
  ];

  doCheck = true;
  preCheck = ''
    export OMP_NUM_THREADS=8
  '';
  checkPhase = "make test";

  meta = with stdenv.lib; {
    description = "Suite of nonlinear differential/algebraic equation solvers";
    homepage    = https://computation.llnl.gov/projects/sundials;
    platforms   = platforms.all;
    maintainers = with maintainers; [ flokli idontgetoutmuch ];
    license     = licenses.bsd3;
  };
}
```

then the tests fail

```
The following tests FAILED:
          8 - ark_KrylovDemo_prec (Failed)
          9 - ark_robertson (Failed)
         10 - ark_robertson_constraints (Failed)
         20 - ark_diurnal_kry_p (Failed)
         21 - ark_diurnal_kry_bbd_p (Failed)
         23 - ark_heat2D (Failed)
         29 - cvDirectDemo_ls (Failed)
         30 - cvDiurnal_kry_bp (Failed)
         31 - cvDiurnal_kry (Failed)
         33 - cvKrylovDemo_ls (Failed)
         36 - cvAdvDiff_diag_p (Failed)
         37 - cvAdvDiff_non_p (Failed)
         38 - cvDiurnal_kry_bbd_p (Failed)
         39 - cvDiurnal_kry_p (Failed)
         40 - cvDiurnal_kry_mpimanyvec (Failed)
         43 - test_nvector_mpi_1000_0 (Failed)
         44 - test_nvector_mpi_4_1000_0 (Failed)
         45 - test_nvector_mpimanyvector_serial_1_1000_100_0 (Failed)
         46 - test_nvector_mpimanyvector_parallel1_4_1000_200_0 (Failed)
         47 - test_nvector_mpimanyvector_parallel2_4_200_1000_0 (Failed)
         48 - test_nvector_mpiplusx_4_1000_9 (Failed)
         89 - test_sunlinsol_spgmr_parallel_100_1_1_50_1e-3_0 (Failed)
         90 - test_sunlinsol_spgmr_parallel_100_1_2_50_1e-3_0 (Failed)
         91 - test_sunlinsol_spgmr_parallel_100_2_1_50_1e-3_0 (Failed)
         92 - test_sunlinsol_spgmr_parallel_100_2_2_50_1e-3_0 (Failed)
         93 - test_sunlinsol_spfgmr_parallel_100_1_50_1e-3_0 (Failed)
         94 - test_sunlinsol_spfgmr_parallel_100_2_50_1e-3_0 (Failed)
         95 - test_sunlinsol_spbcgs_parallel_100_1_50_1e-3_0 (Failed)
         96 - test_sunlinsol_spbcgs_parallel_100_2_50_1e-3_0 (Failed)
         97 - test_sunlinsol_sptfqmr_parallel_100_1_50_1e-3_0 (Failed)
         98 - test_sunlinsol_sptfqmr_parallel_100_2_50_1e-3_0 (Failed)
Errors while running CTest
```

But if I turn off the tests

```
  doCheck = true;
```

and run

```
nix-shell --pure -I nixpkgs=~/nixpkgs test.nix
mpicxx ark_heat2D.cpp -lm -lsundials_arkode -lsundials_nvecparallel -o ark_heat2D
mpiexec -n 4 ./ark_heat2D
```

then I can reproduce the results in the documentation :)

```
2D Heat PDE test problem:
   nprocs = 4
   nx = 60
   ny = 120
   kx = 0.5
   ky = 0.75
   rtol = 1e-05
   atol = 1e-10
   nxl (proc 0) = 30
   nyl (proc 0) = 60

        t      ||u||_rms
   ----------------------
    0.000000    0.000000
    0.015000    0.005780
    0.030000    0.009224
    0.045000    0.011275
    0.060000    0.012497
    0.075000    0.013224
    0.090000    0.013658
    0.105000    0.013916
    0.120000    0.014070
    0.135000    0.014162
    0.150000    0.014216
    0.165000    0.014249
    0.180000    0.014268
    0.195000    0.014280
    0.210000    0.014287
    0.225000    0.014291
    0.240000    0.014293
    0.255000    0.014295
    0.270000    0.014295
    0.285000    0.014296
    0.300000    0.014296
   ----------------------

Final Solver Statistics:
   Internal solver steps = 51 (attempted = 51)
   Total RHS evals:  Fe = 0,  Fi = 513
   Total linear solver setups = 18
   Total linear iterations = 4451
   Total number of Jacobian-vector products = 4451
   Total number of Preconditioner setups = 1
   Total number of Preconditioner solves = 4666
   Total number of linear solver convergence failures = 215
   Total number of Newton iterations = 255
   Total number of nonlinear solver convergence failures = 0
   Total number of error test failures = 0
```

The tests fail on  both macos and ubuntu.

Does anyone have any idea why the tests are failing?

@flokli

**Comment by idontgetoutmuch** [comment]
[log.txt](https://github.com/LLNL/sundials/files/3963576/log.txt)

**Comment by flokli** [comment]
@idontgetoutmuch I wouldn't trust `mpicxx` to not accidentially include anything from your main system into the binary.

**Comment by gardner48** [comment]
Some of the tests (e.g. `ark_KrylovDemo_prec`) are failing because the output varies more than the tolerances used when comparing against the saved output file. This is to be expected with some tests due to differences in compilers, flags, the test machine used, etc. When `SUNDIALS_DEVTESTS=OFF` (the default) the output comparisons are not made and only a subset of tests (those that return a pass/fail status) are run with `make test`. These variations in output seem to be the cause of all the non-MPI test failures. 

As for the MPI tests (e.g., `test_nvector_mpi_1000_0`) it's not clear what the source of the problem is in these tests. In the error log the run command appears to be correct
```
/nix/store/n8jg1860f1wfqyzdr2zfi0r28k6208cx-openmpi-4.0.1/bin/mpirun -n 4 /tmp/nix-build-sundials-5.0.0.drv-0/sundials-5.0.0/build/examples/nvector/parallel/test_nvector_mpi 1000 0
```
and you're able to run an MPI test from the command line. If you test with `SUNDIALS_DEVTESTS=OFF` do the MPI tests (e.g., `test_nvector_mpi_1000_0`) pass?

**Comment by idontgetoutmuch** [comment]
I built with `SUNDIALS_DEVTESTS=OFF` but the same tests fail. It may be that setting `doCheck = true;` in nix somehow overrides this but my nix skills are not good enough to figure this out.

I also tried

```
[nix-shell:~/sundials/examples/nvector/parallel]$ mpicc test_nvector_mpi.c -lm -lsundials_arkode -lsundials_nvecparallel -I.. -o test_nvector_mpi
```

but this failed to link

```
Undefined symbols for architecture x86_64:
  "_SetTiming", referenced from:
      _main in test_nvector_mpi-6644bc.o
  "_Test_N_VAbs", referenced from:
      _main in test_nvector_mpi-6644bc.o
  "_Test_N_VAddConst", referenced from:
      _main in test_nvector_mpi-6644bc.o
```

and I couldn't (easily) see which of the following libraries would resolve this

```
  /nix/store/6ma6vr4dfbkjjh2wf6r7q1f30y3qfq89-sundials-5.0.0/lib:
  total used in directory 3952 available 9223372036849994705
  dr-xr-xr-x  22 root  wheel     704  1 Jan  1970 .
  dr-xr-xr-x   5 root  wheel     160  1 Jan  1970 ..
  -r--r--r--   1 root  wheel  780120  1 Jan  1970 libsundials_arkode.a
  -r--r--r--   1 root  wheel  391816  1 Jan  1970 libsundials_cvode.a
  -r--r--r--   1 root  wheel   41072  1 Jan  1970 libsundials_nvecmanyvector.a
  -r--r--r--   1 root  wheel   48040  1 Jan  1970 libsundials_nvecmpimanyvector.a
  -r--r--r--   1 root  wheel   50536  1 Jan  1970 libsundials_nvecmpiplusx.a
  -r--r--r--   1 root  wheel   89800  1 Jan  1970 libsundials_nvecopenmp.a
  -r--r--r--   1 root  wheel   80360  1 Jan  1970 libsundials_nvecparallel.a
  -r--r--r--   1 root  wheel   74248  1 Jan  1970 libsundials_nvecserial.a
  -r--r--r--   1 root  wheel   17184  1 Jan  1970 libsundials_sunlinsolband.a
  -r--r--r--   1 root  wheel   27432  1 Jan  1970 libsundials_sunlinsoldense.a
  -r--r--r--   1 root  wheel   36648  1 Jan  1970 libsundials_sunlinsolpcg.a
  -r--r--r--   1 root  wheel   38072  1 Jan  1970 libsundials_sunlinsolspbcgs.a
  -r--r--r--   1 root  wheel   39712  1 Jan  1970 libsundials_sunlinsolspfgmr.a
  -r--r--r--   1 root  wheel   40112  1 Jan  1970 libsundials_sunlinsolspgmr.a
  -r--r--r--   1 root  wheel   40392  1 Jan  1970 libsundials_sunlinsolsptfqmr.a
  -r--r--r--   1 root  wheel   32168  1 Jan  1970 libsundials_sunmatrixband.a
  -r--r--r--   1 root  wheel   30272  1 Jan  1970 libsundials_sunmatrixdense.a
  -r--r--r--   1 root  wheel   43104  1 Jan  1970 libsundials_sunmatrixsparse.a
  -r--r--r--   1 root  wheel   45096  1 Jan  1970 libsundials_sunnonlinsolfixedpoint.a
  -r--r--r--   1 root  wheel   39384  1 Jan  1970 libsundials_sunnonlinsolnewton.a
```

From what you (@gardner48) are saying - I don't think it's really worth pursuing this very vigorously at the moment.
